src/borehole_temperature_models/TemperatureModel.py

```diff
@@ -360,11 +202,396 @@ def Compare(
+
+            self._Ts[self._len_zr - 1] = -1.89 - 7.53e-4 * Hi #  type: ignore # Empirically-derived, thickness-dependent seawater freezing point
+
+        # ----------------------------------------------------------------------
+
+        self._LoopImpl(
+            self.t_ungrounding,
+            PreLoopCalc,
+            IntraLoopCallback,
+            calculation_offset=1,
+            update_matrix_index_offset_or_post_loop_callback=self._t_before_ungrounding,
+        )
+
+    # ----------------------------------------------------------------------
+    def Loop4(self) -> None:
+        # ----------------------------------------------------------------------
+        def PreLoopCalc(
+            loop_index: int,
+        ) -> tuple[
+            float,
+            npt.NDArray[np.float64],
+            Optional[npt.NDArray[np.float64]],
+        ]:
+            Hi = self.thkhistory[loop_index + self._t_before_grounding]
+
+            new_zi = np.linspace(0, Hi, self._len_zi)
+            z = np.concatenate((self.zr[:-1], new_zi))
+
+            #  Interpolating previous temperature profile on to the new ice column space domain
+            self._Ts[self._len_zr - 1:] = InterpolatedUnivariateSpline(
+                np.linspace(0, self.thkhistory[loop_index + self._t_before_grounding - 1], self._len_zi),
+                self._Ts[self._len_zr - 1:],
+                k=1,
+            )(new_zi)
+
+            # Grounded ice vertical velocity profile
+
+            dws = self.acchistory[loop_index + self._t_before_grounding] - ((self.thkhistory[loop_index + self._t_before_grounding] - self.thkhistory[loop_index + self._t_before_grounding - 1]) / Constants.dt) # When thickness is variable through time
+
+            return (
+                Hi / self._len_zi,
+                -dws * Lliboutry(new_zi, Hi),
+                z,
+            )
+
+        # ----------------------------------------------------------------------
+        def IntraLoopCallback(
+            loop_index: int,
+        ) -> None:
+            self._Ts[-1] = self.temphistory[loop_index + self._t_before_grounding] # Temperature forcing is constant and equals the first value of the imported Monte-Carlo surface temperature vectors
+
+        # ----------------------------------------------------------------------
+
+        self._LoopImpl(
+            self.t_grounding,
+            PreLoopCalc,
+            IntraLoopCallback,
+            update_matrix_index_offset_or_post_loop_callback=self._t_before_grounding,
+        )
+
+    # ----------------------------------------------------------------------
+    # ----------------------------------------------------------------------
+    # ----------------------------------------------------------------------
+    def _LoopImpl(

```

Comment:
@sashamontelli Can we deep-dive into this function, as I am not entirely sure what it is doing. Thanks!